envision/server.py

```diff
@@ -214,28 +222,40 @@ def sync_run_forever():
         self._thread = threading.Thread(target=sync_run_forever, args=(), daemon=True)
         self._thread.start()
 
-    def _push_frame_to_web_client(self, frame):
+    def _push_frame_to_web_client(self, frames):
         try:
-            self._client.write_message(
-                json.dumps(
-                    {
-                        "state": frame.data,
-                        "current_elapsed_time": frame.timestamp
-                        - self._frames.start_time,
-                        "total_elapsed_time": self._frames.elapsed_time,
-                    }
-                )
-            )
+            frames_formatted = [
+                {
+                    "state": frame.data,
+                    "current_elapsed_time": frame.timestamp - self._frames.start_time,
+                    "total_elapsed_time": self._frames.elapsed_time,
+                }
+                for frame in frames
+            ]
+            self._client.write_message(json.dumps(frames_formatted))
             return False
         except WebSocketClosedError:
             return True
 
-    async def _wait_for_next_frame(self, frame):
+    def _calculate_frame_delay(self, frame_ptr, last_sent_time):
+        if time.time() - last_sent_time > self._timestep_sec:
+            return 0
+        if frame_ptr.next_ is not None:
+            return self._timestep_sec / 10
+        else:
+            # run out of frames to send
+            return 0.99 * self._timestep_sec
+
+    async def _wait_for_next_frame(self, frame_ptr, last_sent_time):
         while True:
-            # TODO: Consider using an asyncio queue instead
-            await asyncio.sleep(self._timestep_sec)
-            if frame.next_ is not None:
-                return frame.next_
+            delay = self._calculate_frame_delay(frame_ptr, last_sent_time)
+            await asyncio.sleep(delay)
+
+            frames_to_send = []
+            while frame_ptr.next_ is not None:
+                frames_to_send.append(frame_ptr)
+                frame_ptr = frame_ptr.next_
```

Comment:
So if `frame_ptr.next_` does not have a `next_`, for example if frame_ptr is the last frame, it would still be stuck in this while loop right? Please feel free to let me know if it is not the case